worker.go

```diff
@@ -55,58 +57,98 @@ func NewWorker(c *Client, wm WorkMap, options ...WorkerOption) *Worker {
 	}
 
 	for _, option := range options {
-		option(&instance)
+		option(&w)
 	}
 
-	if instance.id == "" {
-		instance.id = newID()
+	if w.id == "" {
+		w.id = newID()
 	}
 
-	instance.logger = instance.logger.With(adapter.F("worker-id", instance.id))
+	w.logger = w.logger.With(adapter.F("worker-id", w.id))
 
-	return &instance
+	return &w
 }
 
 // Start pulls jobs off the Worker's queue at its interval. This function runs
 // in its own goroutine, use cancel context to shut it down.
+//
+// Deprecated: use Run instead of Start. Start leaks resources and does not wait
+// for shutdown to complete.
 func (w *Worker) Start(ctx context.Context) error {
-	w.mu.Lock()
-	defer w.mu.Unlock()
+	errc := make(chan error, 1)
+	go func() {
+		errc <- w.runLock(ctx, func(ctx context.Context) error {
+			errc <- nil
+			return w.runLoop(ctx)
+		})
+	}()
+	return <-errc
+}
+
+// Run pulls jobs off the Worker's queue at its interval. This function does
+// not run in its own goroutine so it’s possible to wait for completion. Use
+// context cancellation to shut it down.
+func (w *Worker) Run(ctx context.Context) error {
+	return w.runLock(ctx, w.runLoop)
+}
 
+// runLock runs function f under a run lock. Any attempt to call runLock concurrently
+// will return an error.
+func (w *Worker) runLock(ctx context.Context, f func(ctx context.Context) error) error {
+	w.mu.Lock()
 	if w.running {
+		w.mu.Unlock()
 		return fmt.Errorf("worker[id=%s] is already running", w.id)
 	}
-
 	w.running = true
-	go func() {
-		defer func() {
-			w.running = false
-			w.logger.Info("Worker finished")
-		}()
-
-		for {
-			// Try to work a job
-			if w.WorkOne(ctx) {
-				// Since we just did work, non-blocking check whether we should exit
-				select {
-				case <-ctx.Done():
-					return
-				default:
-					// continue in loop
-				}
-			} else {
-				// No work found, block until exit or timer expires
-				select {
-				case <-ctx.Done():
-					return
-				case <-time.After(w.interval):
-					// continue in loop
-				}
+	w.mu.Unlock()
+
+	defer func() {
+		w.mu.Lock()
+		w.running = false
+		w.mu.Unlock()
+	}()
+
+	return f(ctx)
+}
+
+// runLoop pulls jobs off the Worker's queue at its interval.
+func (w *Worker) runLoop(ctx context.Context) error {
+	defer w.logger.Info("Worker finished")
```

Comment:
please add worker id as a log field

Reply:
That’s already handled on worker initialization 🙂

https://github.com/vgarvardt/gue/blob/ca8abd9787df081da8c085cd0e5fd54dc87c402d/worker.go#L65